viewing_party/party.py

```diff
@@ -1,23 +1,122 @@
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    # an empty watched list is 0.0
+    if not user_data["watched"]:
+        return 0.0
+    # caculate total rating
+    total_rating = 0
+    for movie in user_data["watched"]:
+        total_rating += movie["rating"]
+
+    avg_rating = total_rating / len(user_data["watched"])
 
+    return avg_rating
 
+def get_most_watched_genre(user_data):
+    # an empty list return None
+    if not user_data["watched"]:
+        return None
+    
+    most_genre = {}
+    # bulid a dict for genre count
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre in most_genre:
+            most_genre[genre] +=1
+        else:
+            most_genre[genre] =1
+    # use max to find the most one
+    return max(most_genre, key=most_genre.get)
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    friends_movies = set()
 
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    user_unique = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_movies:
+            user_unique.append(movie)
+    return user_unique
+
+def get_friends_unique_watched(user_data):
+    friends_movies = set()
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    
+    user_watched = {movie["title"] for movie in user_data["watched"]}
+    
+    friends_unique = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched and movie not in friends_unique:
+                friends_unique.append(movie)
+    
+    return friends_unique
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    recommended_movies = []
+    if "subscriptions" in user_data and user_data["subscriptions"]:
+        for movie in  get_friends_unique_watched(user_data):
+            if movie["host"] in user_data["subscriptions"]:
+                recommended_movies.append(movie)
+    return recommended_movies
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    
+    filtered_recommendations = []
+    for movie in  get_available_recs(user_data):
+        if movie["genre"] == get_most_watched_genre(user_data) and get_most_watched_genre(user_data):

```

Comment:
I would consider reversing the order of these comparisons. If `get_most_watched_genre` returns `None`, there's no reason to compare it to `movie["genre"]`. We also shouldn't use the `==` operator if we are comparing something to `None`, which will happen right now if `get_most_watched_genre` returns `None`.

If we check if `get_most_watched_genre(user_data)` exists first, then with python short circuiting if it returns `None`, it will never bother to check movie["genre"] == get_most_watched_genre(user_data).

---------------------------------------------------------

**More about testing for `None` if interested:**
In Python, when we test if a variable is pointing to `None` the PEP8 style guide recommends using `is` or `is not` for comparison rather than operators like `==`. 

`None` is a special object in Python, it’s a `singleton`, meaning there is only ever one instance of `None` in memory. All variables holding `None` are actually pointing to the same instance. By using `is` to compare against `None`, we avoid issues that could be caused by a class’s custom implementation of the comparison operators, and we gain some speed over a check using `==` since we are checking against identity rather than looking up if the variable's held type has a comparison function that should be used.
```python
if movie_title is None or genre is None or rating is None :
    return None
```
If you’d like more info on `is` vs `==` for `None`, here are a couple resources to get started =]
https://peps.python.org/pep-0008/#programming-recommendations
http://jaredgrubb.blogspot.com/2009/04/python-is-none-vs-none.html